### PR TITLE
DAML-LF: generic map values protobuf format

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
This PR advances the state of Generic Maps #2256.

In this PR, we set up the serialization of GenMap inside the ledger.

### Pull Request Checklist

- [X] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [X] Include appropriate tests
- [X] Set a descriptive title and thorough description
- [X] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags, if appropriate
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/spec/value.rst [hunk 1]
```diff
@@ -579,3 +594,34 @@ In this version, these fields are included:
...
+* `message Value`_ value
+
+Both ``key`` and ``value`` are required.
+
+message GenMap
+^^^^^^^^^^^
+
+*since version 7*
+
+A map where keys are values are homogeneous.
```
Comment (gerolf-da) [inline daml-lf/spec/value.rst]:
```suggestion
A map where keys and values are homogeneous.
```

File: daml-lf/spec/value.rst [hunk 2]
```diff
@@ -579,3 +594,34 @@ In this version, these fields are included:
...
+
+In this version, these fields are included:
+
+* repeated `message GenMap.Entry`_ entries
+
+The ``keys`` and the ``value`` fields of every member of ``entries`` must
+conform to the same types. Entries occur in assertion order, if two ore more
```
Comment (gerolf-da) [inline daml-lf/spec/value.rst]:
```suggestion
conform to the same types. Entries occur in insertion order. If two ore more
```

File: daml-lf/spec/value.rst [hunk 3]
```diff
@@ -579,3 +594,34 @@ In this version, these fields are included:
...
+entries have the same keys, the last one override the formers.
```
Comment (gerolf-da) [inline daml-lf/spec/value.rst]:
```suggestion
entries have the same keys, the last one override the former entries.
```

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/VersionTimeline.scala [hunk 4]
```diff
@@ -55,6 +55,7 @@ private[digitalasset] object VersionTimeline {
       This(That(TransactionVersion("8"))),
       Both(This(ValueVersion("5")), LanguageVersion(LMV.V1, "6")),
       Both(This(ValueVersion("6")), LanguageVersion(LMV.V1, "7")),
+      Both(This(ValueVersion("7")), LanguageVersion(LMV.V1, "dev")),
```
Comment (gerolf-da) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/VersionTimeline.scala]: Does it need the LanguageVersion "dev" here?

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/value/ValueCoder.scala [hunk 5]
```diff
@@ -515,9 +521,19 @@ object ValueCoder {
...
-            // FIXME https://github.com/digital-asset/daml/issues/2256
-            throw Err("Trying to serialize GenMap which are not currently not supported.")
+          case ValueGenMap(entries) =>
+            val protoMap = proto.GenMap.newBuilder()
+            entries.foreach {
+              case (key, value) =>
+                protoMap.addEntries(
+                  proto.GenMap.Entry
+                    .newBuilder()
+                    .setKey(go(nesting, key))
```
Comment (gerolf-da) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/value/ValueCoder.scala]: Does this need `newNesting` as well here?

Comment (remyhaemmerle-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/value/ValueCoder.scala]: fixed. Thanks

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/VersionTimeline.scala (hunk 4 above)
Comment (remyhaemmerle-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/VersionTimeline.scala]: I do not want that somebody insert a line after this one before we freeze 1.8.
I add a comment about that.

Comment (gerolf-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/VersionTimeline.scala]: 👍

File: daml-lf/transaction-scalacheck/src/main/scala/com/digitalasset/daml/lf/value/ValueGenerators.scala [hunk 6]
```diff
@@ -196,6 +196,14 @@ object ValueGenerators {
     } yield ValueTextMap[ContractId](SortedLookupList(Map(list: _*)))
   def valueMapGen: Gen[ValueTextMap[ContractId]] = valueMapGen(0)
 
+  private def valueGenMapGen(nesting: Int) =
+    for {
+      list <- Gen.listOf(for {
+        k <- Gen.lzy(valueGen(nesting)); v <- Gen.lzy(valueGen(nesting))
```
Comment (S11001001) [inline daml-lf/transaction-scalacheck/src/main/scala/com/digitalasset/daml/lf/value/ValueGenerators.scala]:
```suggestion
        Gen.zip(Gen.lzy(valueGen(nesting)), Gen.lzy(valueGen(nesting)))
```

Without the `for` trappings.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/VersionTimeline.scala [hunk 7]
```diff
@@ -55,6 +55,10 @@ private[digitalasset] object VersionTimeline {
       This(That(TransactionVersion("8"))),
       Both(This(ValueVersion("5")), LanguageVersion(LMV.V1, "6")),
       Both(This(ValueVersion("6")), LanguageVersion(LMV.V1, "7")),
+      // FIXME https://github.com/digital-asset/daml/issues/2256
+      //  * change the following line when LF 1.8 is frozen.
+      //  * do not insert line after this once until 1.8 is frozen.
+      Both(This(ValueVersion("7")), LanguageVersion(LMV.V1, Dev)),
```
Comment (S11001001) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/VersionTimeline.scala]:
```suggestion
      This(This(ValueVersion("7"))),
```

File: daml-lf/spec/value.rst [hunk 8]
```diff
@@ -579,3 +594,34 @@ In this version, these fields are included:
...
+As of version 7, these fields are included:
+
+* `message Value`_  key
+
+* `message Value`_ value
+
+Both ``key`` and ``value`` are required.
+
+message GenMap
+^^^^^^^^^^^
```
Comment (remyhaemmerle-da) [inline daml-lf/spec/value.rst]:
```suggestion
^^^^^^^^^^^^^^
```

File: daml-lf/transaction-scalacheck/src/main/scala/com/digitalasset/daml/lf/value/ValueGenerators.scala [hunk 9]
```diff
@@ -196,6 +196,13 @@ object ValueGenerators {
     } yield ValueTextMap[ContractId](SortedLookupList(Map(list: _*)))
   def valueMapGen: Gen[ValueTextMap[ContractId]] = valueMapGen(0)
 
+  private def valueGenMapGen(nesting: Int) =
+    Gen
+      .listOf(Gen.zip(Gen.lzy(valueGen(nesting)), Gen.lzy(valueGen(nesting))))
+      .flatMap(list => ValueGenMap[ContractId](ImmArray(list)))
```
Comment (S11001001) [inline daml-lf/transaction-scalacheck/src/main/scala/com/digitalasset/daml/lf/value/ValueGenerators.scala]:
```suggestion
      .map(list => ValueGenMap[ContractId](ImmArray(list)))
```

I'll leave you to 😨 😢 about why it compiled

**Pull request merged**